src/mcts/node.h

```diff
@@ -28,40 +29,116 @@
 
 namespace lczero {
 
+// Children of a node are stored the following way:
+// * Edges and Nodes edges point to are stored separately.
+// * There may be dangling edges (which don't yet point to any Node object yet)
+// * Edges are stored are a simple array on heap.
+// * Nodes are stored as a linked list, and contain index_ field which shows
+//   which edge of a parent that node points to.
+//
+// Example:
+//                                Parent Node
+//                                    |
+//        +-------------+-------------+----------------+--------------+
+//        |              |            |                |              |
+//   Edge 0(Nf3)    Edge 1(Bc5)     Edge 2(a4)     Edge 3(Qxf7)    Edge 4(a3)
+//    (dangling)         |           (dangling)        |           (dangling)
+//                   Node, Q=0.5                    Node, Q=-0.2
+//
+//  Is represented as:
+// +--------------+
+// | Parent Node  |
+// +--------------+                                        +--------+
+// | edges_       | -------------------------------------> | Edge[] |
+// |              |    +------------+                      +--------+
+// | child_       | -> | Node       |                      | Nf3    |
+// +--------------+    +------------+                      | Bc5    |
+//                     | index_ = 1 |                      | a4     |
+//                     | q_ = 0.5   |    +------------+    | Qxf7   |
+//                     | sibling_   | -> | Node       |    | a3     |
+//                     +------------+    +------------+    +--------+
+//                                       | index_ = 3 |
+//                                       | q_ = -0.2  |
+//                                       | sibling_   | -> nullptr
+//                                       +------------+
+
 class Node;
-class Node_Iterator {
+class Edge {
  public:
-  Node_Iterator(Node* node) : node_(node) {}
-  Node* operator*() { return node_; }
-  Node* operator->() { return node_; }
-  bool operator==(Node_Iterator& other) { return node_ == other.node_; }
-  bool operator!=(Node_Iterator& other) { return node_ != other.node_; }
-  void operator++();
+  Edge(Move move) : move_(move) {}
```

Comment:
For small objects like this (4 bytes, or later 2 bytes) it's actually better to pass by value rather than reference (which is 8 bytes by itself plus adds a level of indirection)